lambda/translquote.ml

```diff
@@ -0,0 +1,3144 @@
+ * The above copyright notice and this permission notice shall be included    *
+ * in all copies or substantial portions of the Software.                     *
+ *                                                                            *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR *
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,   *
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL    *
+ * THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER *
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING    *
+ * FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER        *
+ * DEALINGS IN THE SOFTWARE.                                                  *
+ ******************************************************************************)
+
+open Asttypes
+open Lambda
+open Misc
+open Typedtree
+open Types
+open Debuginfo.Scoped_location
+open Longident
+
+let camlinternalQuote =
+  lazy
+    (match
+       Env.open_pers_signature "CamlinternalQuote" (Lazy.force Env.initial)
+     with
+    | exception Not_found -> fatal_error "Module CamlinternalQuote unavailable."
+    | path, _, env -> path, env)
+
+let combinator modname field =
+  lazy
+    (let _, env = Lazy.force camlinternalQuote in
+     let lid =
+       match unflatten (String.split_on_char '.' modname) with
+       | None -> Lident field
+       | Some lid -> Ldot (lid, field)
+     in
+     match Env.find_value_by_name_lazy lid env with
+     | p, _ -> transl_value_path Loc_unknown env p
+     | exception Not_found ->
+       fatal_error
+         ("Primitive CamlinternalQuote." ^ modname ^ "." ^ field ^ " not found."))
+
+(* Strings and booleans *)
+
+let string loc s = Lconst (Const_base (Const_string (s, loc, None)))
+
+let true_ = Lconst (Const_base (Const_int 1))
+
+let false_ = Lconst (Const_base (Const_int 0))
+
+let quote_bool b = if b then true_ else false_
+
+(* Lambdas that represent the option type *)
+
+let none = Lconst (Const_base (Const_int 0))
+
+let some x =
+  Lprim (Pmakeblock (0, Immutable, None, alloc_heap), [x], Loc_unknown)
+
+let option opt = match opt with None -> none | Some x -> some x
+
+let string_option loc s = option (Option.map (string loc) s)
+
+(* Lambdas for lists and tuples *)
+
+let nil = Lconst (Const_base (Const_int 0))
+
+let cons hd tl =
+  Lprim (Pmakeblock (0, Immutable, None, alloc_heap), [hd; tl], Loc_unknown)
+
+let hd l = Lprim (Pfield (0, Immediate, Reads_vary), [l], Loc_unknown)
+
+let tl l = Lprim (Pfield (1, Immediate, Reads_vary), [l], Loc_unknown)
+
+let rec mk_list list =
+  match list with [] -> nil | hd :: tl -> cons hd (mk_list tl)
+
+let pair (x, y) =
+  Lprim (Pmakeblock (0, Immutable, None, alloc_heap), [x; y], Loc_unknown)
+
+let triple (x, y, z) =
+  Lprim (Pmakeblock (0, Immutable, None, alloc_heap), [x; y; z], Loc_unknown)
+
+(* Let-expressions *)
+
+let bind id def body =
```

Comment:
This top section makes me somewhat nervous, it makes a lot of decisions about how these things are encoded in lambda which feel brittle and not always correct.